python/src/otel/otel_sdk/opentelemetry/instrumentation/aws_lambda/__init__.py

```diff
@@ -99,36 +96,32 @@ def _uninstrument(self, **kwargs):
 
 def _instrument(tracer, wrapped_module_name, wrapped_function_name):
     def _instrumented_lambda_handler_call(call_wrapped, instance, args, kwargs):
-        orig_handler_name = ".".join(
-            [wrapped_module_name, wrapped_function_name]
+        lambda_event, lambda_context = args[0], args[1]
+
+        parent_context = get_global_textmap().extract(
+            lambda_event.get("headers", {}).update(
```

Comment:
I don't think we know headers will always be located at lambda_event.headers. If a spec guaranteed all lambda entry points follow this convention we could do this, but I don't believe such a spec exists. See how eventContextExtractor in [JS Instrumentation](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/68e54490b207980b080d6642be48e81d17659cc7/plugins/node/opentelemetry-instrumentation-aws-lambda#aws-lambda-instrumentation-options) handles this by requiring user to provide header extraction function. Note JS requires user to fully extract into context which is unnecessary for here since you are using global propagator to extract. Here could require user to just return headers from event.

Reply:
Hm I see what you're saying. Yeah [this guide I found online](https://aws.amazon.com/premiumsupport/knowledge-center/custom-headers-api-gateway-lambda/) gives an example on how to populate `"headers"` in API Gateway, but it's not a general solution.

I think we could do something like JS for the manual instrumentation case, but not the auto instrumentation case. In auto instrumentation the `AwsLambdaInstrumentor` will run without hooks and later on it will be too late to add the hooks.

So either we force users to manual instrument `AwsLambdaInstrumentor` in their code before their `lambda_handler`, or we define a mapping "ourselves" here as `events["headers"]` and explain to users how they can add them...

I think we can add the hooks in a subsequent PR, the question here is just do we want to give users the expectation that we will looks for HTTP headers in the `event["headers"]`? I think it could for us even though there is no spec, as long as we document it somewhere here?

Reply:
So `event["headers"]` may or may not be pointing into the actual customer event. If you invoke lambda directly (cli/sdk) `event` is your custom event type so yes we could point to one place the header must be, however, if lambda is invoked via some other AWS service (API Gateway, SQS, etc) `event` is some AWS specified type and your custom event is within that type somewhere.

Reply:
First of all, many thanks for looking into this.

My 2c:

Lambda handlers always accept 2 arguments: event and context. Event is the message delivered by a `Carrier` (API GW, SQS, etc) and should have both, data and metadata about the message that can include the context. [Propagators](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) define the rules about how the context should be propagated with a particular `Carrier`. 

So, it should be safe to assume that `event` should contain the required metadata with the context. As the default implementation, it should be safe to assume that `event["headers"]` will contain the context data and it may be 95% of all cases where people use API-GW and SQS triggered lambdas. In other cases, the users can provide their own extractor and injector implementations, as JS lambda implementation does.

In the case where there is no data in the message (i.e. `event["header"]`) that carries the context for the defined propagator, it should default to the span with no parent and start a new trace, as opposed to defaulting to AWS X-Ray propagator. 

Also, specifying AWS X-Ray propagator should be a choice, i.e. the user should explicitly define it in `OTEL_PROPAGATORS=aws_xray` env variable.

Hope it makes sense.

Reply:
@garrettwegan

I guess we need to find out the different branching paths for what an "event could be". For now, I followed the spec where it says [how to determine the parent span in Lambda](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md#determining-the-parent-of-a-span) to guide a "**guess**" for what API Gateway would look like, but I haven't tested it so I wouldn't be comfortable merging this _just_ yet. But if we test it and it works (In the X-Ray case, the not X-Ray case, and the not _any_ propagator case), and people agree with it, we can make this incremental improvement.

@michaelhyatt in trying to follow this spec on [how to determine the parent span in Lambda](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md#determining-the-parent-of-a-span) I made several updates to this PR. It still requires serious testing though:
* test with API Gateway configured in its 2 possible forms
* test when multiple combinations of propagators are set via `OTEL_PROPAGATORS`

I can't test it this week but if we get a chance to test it in the near future and people are okay with checking `event["headers"]` as a default we can probably make this incremental improvement to unblock you.